lib/api/model/model.dart

```diff
@@ -698,3 +710,83 @@ class DmMessage extends Message {
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
 }
+
+enum MessageEditState {
+  none,
+  edited,
+  moved;
+
+  // Adapted from the shared code:
+  //   https://github.com/zulip/zulip/blob/1fac99733/web/shared/src/resolved_topic.ts
+  // Pattern for an arbitrary resolved-topic prefix.
+  // These always begin with the canonical prefix, but can go on longer.
+  // It's designed to remove a weird "✔ ✔✔ " prefix, if present.
+  static RegExp _resolvedTopicPrefixRe = RegExp('^✔ [ ✔]*');
+
+  static bool _isResovled(String topic) {
+    return _resolvedTopicPrefixRe.matchAsPrefix(topic) != null;
+  }
+
+  /// Whether two topics are equal, ignoring any resolved-topic prefix.
+  ///
+  /// When a topic is resolved, the clients agree on adding a ✔ prefix to the
+  /// topic string. Topics whose only difference is the ✔ prefix are considered
+  /// the same. This helper can be helpful when checking if a message has been
+  /// moved.
+  static bool areSameTopic(String topic, String prevTopic) {
```

Comment:
```suggestion
  static bool sameTopicModuloResolve(String topic, String prevTopic) {
```

That makes it a bit clearer what this method's peculiarity is. For most purposes, if we have the resolved and unresolved version of a topic coexisting, they're distinct topics just the same as if the names were unrelated.